### ISSUE TITLE
ES REST client does not support HTTPS

### ISSUE THREAD

**Issue by ngrigoriev** [JanusGraph/janusgraph#645]
In ElasticSearchSetup class I see this:

```
for (String host : config.get(INDEX_HOSTS)) {
                String[] hostparts = host.split(":");
                String hostname = hostparts[0];
                int hostport = defaultPort;
                if (hostparts.length == 2) hostport = Integer.parseInt(hostparts[1]);
                log.debug("Configured remote host: {} : {}", hostname, hostport);
                hosts.add(new HttpHost(hostname, hostport, "http"));
            }
```

I think we should have an option for HTTPS as well.

**Comment by sjudeng** [comment]
Included in #612

**Comment by ngrigoriev** [comment]
Well, while I do agree that it would be included in #612, the fix for this one would be literally 2 or 3 lines (just support for HTTPS). I have actually tested it and it worked fine with AWS ES 5.5. managed service. And #612 requires some kind of authentication, which is another problem and it is outside of the scope of HttpHost. Unless a PR is coming for #612, I would probably recommend addressing these problems separately.

**Comment by sjudeng** [comment]
Adding tests might be the bulk of the work here (and in #612). If it were me I'd do both SSL and authentication together but we can reopen this in case you or anyone want to try and add tests and implementation to support HTTPS separately.

**Comment by amcp** [comment]
I like @sjudeng idea to put both in the same PR

**Comment by ngrigoriev** [comment]
I would be very happy to provide first the PR for this one (we need it like now :) ) and then for #612.

**Comment by mbrukman** [comment]
@ngrigoriev — in parallel with developing your patch, please also [sign the CLA][cla] and configure your git client accordingly. You may need to have your company sign a CCLA rather than signing an ICLA; please check with your legal department. Looking forward to your contributions!

[cla]: https://github.com/JanusGraph/janusgraph/blob/master/CONTRIBUTING.md

**Comment by davidclement90** [comment]
How you plan to test your feature without using X-pack/shield ?

**Comment by ngrigoriev** [comment]
I did not explore yet the Shield part. I think am done with just plain HTTPS support. I tested it against AWS ES service, which only allows HTTPS. As for Shield, I was thinking about an HTTPS proxy with Basic authentication. It seems that's what X-Pack/Shield is using.

Will deal with CLA this morning.

**Comment by ngrigoriev** [comment]
I have just realized that I need one more thing to have in this PR :) AWS request authentication. I mean - in addition to HTTP Basic authentication needed for Shield/X-Pack I also need AWS authentication to be able to sign the requests for AWS ES service. This goes beyond the scope of both this issue and #612. What would you propose? Extend #612 or open another issue and link all together? I am developing the change that would support both auth methods (and others if needed)

**Comment by sjudeng** [comment]
I think we would want to avoid adding custom code (and the AWS SDK dependency) to support AWS-specific auth directly in janusgraph-es. My concern is it would be difficult to test and maintain the AWS client implementation. What would make sense to me though would be if we wanted to look at refactoring to expose higher-level abstractions to allow users to accomplish this on their end. For example maybe [RestElasticSearchClient](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java) could be updated to support a plugable authentication mechanism or support could be added for custom user [ElasticSearchClient](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchClient.java) implementations. Then the AWS example could be provided in documentation and/or [janusgraph-examples](https://github.com/JanusGraph/janusgraph/tree/master/janusgraph-es).

What do others think?

**Comment by ngrigoriev** [comment]
The implementation turns out to be, in fact, about 5 lines of code. But it does require a dependency on aws-java-sdk-core - to implement the V4 signing. The required hook is already available in the RestClientBuilder. So it does not really require any modifications/extensions of ElasticSearchClient. However, dependency is a dependency....

Technically this signer requires no parameters. So, theoretically, it can be done in a way similar to how DynamoDB support is done for storage:

```
storage.dynamodb.client.credentials.class-name=com.amazonaws.auth.BasicAWSCredentials
storage.dynamodb.client.credentials.constructor-args=fake,fake
```

So, it could be like a custom authenticator implemented as HttpRequestInterceptor wrapped into some something. So it would get just "constructor args".

**Comment by ngrigoriev** [comment]
Yes, this would keep AWS dependency out, would allow (potentially) to develop something else similar to AWS auth. And it is not exactly cast in stone. If more external customization is ever needed, additional interfaces can be added to that "auth class" and discovered at runtime.

Oh, yes, absolutely, under "elasticsearch", this is ES-specific - I did not mean that config example to be exactly like this.

Ok, agreed, will go this route with an externalized/configurable interceptor and a separate mini-project implementing that thing for AWS.

**Comment by ngrigoriev** [comment]
The custom AWS "authenticator" would look like https://github.com/ngrigoriev/janusgraph-es-aws-authenticator
